internal/backend/remote-state/kubernetes/client.go

```diff
@@ -90,46 +120,94 @@ func (c *RemoteClient) Put(data []byte) error {
 		return err
 	}
 
-	secret, err := c.getSecret(secretName)
+	chunks := chunkPayload(payload, defaultChunkSize)
+	existingSecrets, err := c.getSecrets()
 	if err != nil {
-		if !k8serrors.IsNotFound(err) {
+		return err
+	}
+
+	for idx, data := range chunks {
+		secretName, err := c.createSecretName(idx)
+		if err != nil {
 			return err
 		}
 
-		secret = &unstructured.Unstructured{
-			Object: map[string]interface{}{
-				"metadata": metav1.ObjectMeta{
-					Name:        secretName,
-					Namespace:   c.namespace,
-					Labels:      c.getLabels(),
-					Annotations: map[string]string{"encoding": "gzip"},
+		secret, err := c.getSecret(secretName)
+		if err != nil {
+			if !k8serrors.IsNotFound(err) {
+				return err
+			}
+			secret = &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"metadata": metav1.ObjectMeta{
+						Name:        secretName,
+						Namespace:   c.namespace,
+						Labels:      c.getLabels(),
+						Annotations: map[string]string{"encoding": "gzip"},
+					},
 				},
-			},
+			}
+			secret, err = c.kubernetesSecretClient.Create(ctx, secret, metav1.CreateOptions{})
+			if err != nil {
+				return err
+			}
 		}
 
-		secret, err = c.kubernetesSecretClient.Create(ctx, secret, metav1.CreateOptions{})
+		setState(secret, data)
+		_, err = c.kubernetesSecretClient.Update(ctx, secret, metav1.UpdateOptions{})
 		if err != nil {
 			return err
 		}
 	}
 
-	setState(secret, payload)
-	_, err = c.kubernetesSecretClient.Update(ctx, secret, metav1.UpdateOptions{})
-	return err
+	// remove old secrets
+	existingSecretCount := len(existingSecrets)
+	newSecretCount := len(chunks)
+	if existingSecretCount == newSecretCount {
+		return nil
+	}
+	for i := newSecretCount; i < existingSecretCount; i++ {
+		err := c.deleteSecret(fmt.Sprintf("%s-part-%d", secretName, i))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// chunkPayload splits the state payload into byte arrays of the given size
+func chunkPayload(buf []byte, size int) [][]byte {
+	chunks := make([][]byte, 0, len(buf)/size+1)
+	for len(buf) >= size {
+		var chunk []byte
+		chunk, buf = buf[:size], buf[size:]
+		chunks = append(chunks, chunk)
+	}
+	if len(buf) > 0 {
+		chunks = append(chunks, buf[:len(buf)])
```

Comment:
Is this not the same as `chunks = append(chunks, buf)`? 
Unsure what the purpose of the slice indexing is...

Reply:
Yep this was unneeded – removed.